radixdlt/src/main/java/com/radixdlt/consensus/liveness/WeightedRotatingLeaders.java

```diff
@@ -0,0 +1,160 @@
+
+	public WeightedRotatingLeaders(ValidatorSet validatorSet, Comparator<Validator> comparator, int cacheSize) {
+		this.validatorSet = validatorSet;
+		this.weightsComparator = Comparator
+			.comparing(Entry<Validator, UInt256>::getValue)
+			.thenComparing(Entry::getKey, comparator);
+		this.nextLeaderComputer = new CachingNextLeaderComputer(validatorSet, weightsComparator, cacheSize);
+	}
+
+	private static class CachingNextLeaderComputer {
+		private final ValidatorSet validatorSet;
+		private final Comparator<Entry<Validator, UInt256>> weightsComparator;
+		private final Map<Validator, UInt256> weights;
+		private final Validator[] cache;
+		private final UInt128[] powerArray;
+		private View curView;
+
+		private CachingNextLeaderComputer(ValidatorSet validatorSet, Comparator<Entry<Validator, UInt256>> weightsComparator, int cacheSize) {
+			this.validatorSet = validatorSet;
+			this.weightsComparator = weightsComparator;
+			this.weights = new HashMap<>();
+			this.cache = new Validator[cacheSize];
+			this.powerArray = validatorSet.getValidators().stream().map(Validator::getPower).toArray(UInt128[]::new);
+			this.resetToView(View.of(0));
+		}
+
+		private Validator computeHeaviest() {
+			final Entry<Validator, UInt256> max = weights.entrySet().stream()
+				.max(weightsComparator)
+				.orElseThrow(() -> new IllegalStateException("Weights cannot be empty"));
+			return max.getKey();
+		}
+
+		private void computeNext() {
+			// Reset current leader by subtracting total power
+			final int curIndex = (int) (this.curView.number() % cache.length);
+			final Validator curLeader = cache[curIndex];
+			weights.merge(curLeader, validatorSet.getTotalPower(), UInt256::subtract);
+
+			// Add weights relative to each validator's power
+			for (Validator validator : validatorSet.getValidators()) {
+				weights.merge(validator, UInt256.from(validator.getPower()), UInt256::add);
+			}
+
+			// Compute next leader by getting heaviest validator
+			this.curView = this.curView.next();
+			int index = (int) (this.curView.number() % cache.length);
+			cache[index] = computeHeaviest();
+		}
+
+		private Validator checkCacheForProposer(View view) {
+			if (view.compareTo(curView) <= 0 && view.number() > curView.number() - cache.length) {
+				return cache[(int) (view.number() % cache.length)];
+			}
+
+			return null;
+		}
+
+		private void computeToView(View view) {
+			while (view.compareTo(curView) > 0) {
+				computeNext();
+			}
+		}
+
+		private Validator resetToView(View view) {
+			// reset if view isn't in cache
+			if (curView == null || view.number() < curView.number() - cache.length) {
+				// after cappedLCM is executed, the following invariant will be true:
+				// (lcm > 0 && lcm < 2^64) || lcm == null
+				// This is due to use of 2^64 cap and also the invariant from ValidatorSet
+				// that powerArray will always be non-zero
+				UInt128 lcm = MathUtils.cappedLCM(UInt128.from(view.number()), powerArray);
```

Comment:
LCM is constant for validator set, so should be precomputed.  Can be capped at Long.MAX_VALUE.